packages/flutter/lib/src/widgets/text.dart

```diff
@@ -671,6 +671,7 @@ class Text extends StatelessWidget {
       result = MouseRegion(
         cursor: DefaultSelectionStyle.of(context).mouseCursor ?? SystemMouseCursors.text,
         child: _SelectableTextContainer(
+          selectableId: key,
```

Comment:
Instead of `key`, what if you pass `this` as the selectableId? Then use only a single SelectionListener and:

  1. Walk the tree until you find where `selectionDetails.ranges.first.selectableId == textWidget`.
  2. Color that subtree red given the matching SelectedContentRange. I think you could do that with a static recursive method like `_makeRed` below, which might help break up the size of your logic in _colorSelectionRed.
  3. If there are still more ranges in `selectionDetails.ranges`, continue walking the tree with the new range.
  4. Repeat until you've called `_makeRed` on all the ranges.
  5. Return your new tree, which is the old one but with the selected part red.

I'm assuming that selectionDeteails.ranges is ordered in traversal order.

```dart
static InlineSpan _makeRed(TextSpan textSpan, SelectedContentRange range) {
  assert(textSpan == range.selectableId);
  // I leave the rest as an exercise for the reader.
}
```

Reply:
Benefits, if it works:

 * No explicit selectionId needed on Text.
 * Only one SelectionListener is needed for your examples.
 * No separate data structure needed in your examples (`dataSourceMap` etc.).